### PR TITLE
Add logging into e2e framework

### PR DESCRIPTION
**Pull request by vcastellm**
# Description

This PR enable optionally logging to different io.Writers, by default it saves the e2e logs output to a generated directory and upload the logs as an artifact for later inspection.

# Changes include

- [ ] Bugfix (non-breaking change that solves an issue)
- [ ] Hotfix (change that solves an urgent issue, and requires immediate attention)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (change that is not backwards-compatible and/or changes current functionality)

# Checklist

- [x] I have assigned this PR to myself
- [x] I have added at least 1 reviewer
- [x] I have added the relevant labels
- [ ] I have updated the official documentation
- [x] I have added sufficient documentation in code

## Testing

- [x] I have tested this code with the official test suite
- [x] I have tested this code manually

### Manual tests

Run in local using `make test-e2e`

# Additional comments

This PR should be merged after https://github.com/0xPolygon/polygon-edge/pull/769

### CODE REVIEW

File: e2e/framework/testserver.go [hunk 1]
```diff
@@ -452,14 +448,12 @@ func (t *TestServer) SwitchIBFTType(typ fork.IBFTType, from uint64, to, deployme
...
-	if t.Config.ShowsLog {
-		stdout := io.Writer(os.Stdout)
-		t.cmd.Stdout = stdout
-		t.cmd.Stderr = stdout
-	}
+	stdout := t.GetStdout()
+	t.cmd.Stdout = stdout
+	t.cmd.Stderr = stdout
 
 	return t.cmd.Run()
```
Comment (Kourin1996) [inline e2e/framework/testserver.go:458]: I have one suggestion. Right now, the logs of all e2e test are saved to the files separated by server index. It's hard to know what test the log is outputted from. Is it possible to write header including test information every time before running command in each e2e test?

**Review by Kourin1996 — COMMENTED** [review]
Thank you for PR!

File: e2e/framework/testserver.go (hunk 1 above)
Comment (zivkovicmilos) [reply e2e/framework/testserver.go:458]: Love this idea ➕

**Review by zivkovicmilos — APPROVED** [review]
Love this improvement 💯

**Comment by ferranbt** [comment]
Not sure if I follow the format of the output logs. As was described in the task, each test has a name (i.e. QuorumNodesDropNodeAndSyncBack) and the format of the output directory looks like this:
```
./e2e-tests-<timestamp1>/QuorumNodesDropNodeAndSyncBack/node1.log
./e2e-tests-<timestamp1>/QuorumNodesDropNodeAndSyncBack/node2.log
./e2e-tests-<timestamp1>/QuorumNodesDropNodeAndSyncBack/node3.log
./e2e-tests-<timestamp1>/<test_name_2>/node3.log
```
Then, it is easy to organize the outputs by time (all the outputs in the same `timestamp` folder belong to the same execution) and by a specific test.

File: .github/workflows/e2e.yaml [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+      - uses: actions/checkout@v2
+      - name: Install Go
+        uses: actions/setup-go@v2
+        with:
+          go-version: 1.17
+      - name: Run tests
+        run: make test-e2e
+      - name: Archive test logs
+        if: always()
+        uses: actions/upload-artifact@v3
```
Comment (ZeljkoBenovic) [inline .github/workflows/e2e.yaml:26]: Should we set retention time? Maybe delete log artifacts older than 30 days ?

**Review by ZeljkoBenovic — APPROVED** [review]
LGTM 💯 

This will help a lot with troubleshooting process 👍

**Comment by vcastellm** [comment]
> Not sure if I follow the format of the output logs. As was described in the task, each test has a name (i.e. QuorumNodesDropNodeAndSyncBack) and the format of the output directory looks like this:
> 
> ```
> ./e2e-tests-<timestamp1>/QuorumNodesDropNodeAndSyncBack/node1.log
> ./e2e-tests-<timestamp1>/QuorumNodesDropNodeAndSyncBack/node2.log
> ./e2e-tests-<timestamp1>/QuorumNodesDropNodeAndSyncBack/node3.log
> ./e2e-tests-<timestamp1>/<test_name_2>/node3.log
> ```
> 
> Then, it is easy to organize the outputs by time (all the outputs in the same `timestamp` folder belong to the same execution) and by a specific test.

Totally right, I removed the detailed folder naming during the refactor and forgot to add it again, now it looks like:

image

File: .github/workflows/e2e.yaml (hunk 2 above)
Comment (vcastellm) [reply .github/workflows/e2e.yaml:26]: Didn't know that was possible :), 30 looks like more than enough, added!

File: e2e/framework/testserver.go (hunk 1 above)
Comment (vcastellm) [reply e2e/framework/testserver.go:458]: Better with this? https://github.com/0xPolygon/polygon-edge/pull/771#issuecomment-1262192585

**Review by Kourin1996 — APPROVED** [review]

**Pull request merged**